components/suggest/src/db.rs

```diff
@@ -270,13 +270,20 @@ impl<'a> SuggestDao<'a> {
         let suggestions_limit = &query.limit.unwrap_or(-1);
         let suggestions = self.conn.query_rows_and_then_cached(
             "SELECT s.id, k.rank, s.title, s.url, s.provider, s.score, k.confidence, k.keyword_suffix
-                     FROM suggestions s
-                     JOIN prefix_keywords k ON k.suggestion_id = s.id
-                     WHERE k.keyword_prefix = :keyword_prefix AND s.provider = :provider
-                     ORDER by s.score DESC
-                     LIMIT :suggestions_limit",
+                FROM suggestions s
+                JOIN prefix_keywords k ON k.suggestion_id = s.id
+                JOIN (
+                SELECT
+                    suggestion_id,
+                    MAX(rank) as highest_rank
+                    FROM prefix_keywords WHERE keyword_prefix = :keyword_prefix AND keyword_suffix LIKE :keyword_suffix || '%' GROUP BY suggestion_id
+                ) h ON k.suggestion_id = h.suggestion_id AND k.rank = h.highest_rank
+                WHERE k.keyword_prefix = :keyword_prefix AND k.keyword_suffix LIKE :keyword_suffix || '%' AND s.provider = :provider
+                ORDER BY s.score DESC
+                LIMIT :suggestions_limit",
```

Comment:
Nit: Can you add a newline before the `GROUP BY`?

Reply:
Now that you're joining on the inner `SELECT`, do you still need the outer `JOIN prefix_keywords k ON k.suggestion_id = s.id` and `WHERE k.keyword_prefix = :keyword_prefix AND k.keyword_suffix LIKE :keyword_suffix || '%' AND s.provider = :provider`?

Bigger picture: is there a way to test performance?  Adding the nested query seems slightly scary to me, it would be nice if there was a performance suite that would allow us to say this stayed within X% of the current preformance.

Reply:
Hi @tiftran, we don't use full keywords for AMO. We do use them for Pocket -- [desktop picks the longest full keyword that starts with the query](https://searchfox.org/mozilla-central/rev/34ba7989ae53b9112d543ce39d350b79dc2a16b4/browser/components/urlbar/private/PocketSuggestions.sys.mjs#90-103) -- but since there are no plans to enable Pocket again, IMO there's no reason to worry about that right now.

Reply:
To @bendk's point, this can be written more simply (GitHub won't let me suggest a change for deleted lines, so I'll just paste the whole query here):

```
            "SELECT s.id, MAX(k.rank) AS rank, s.title, s.url, s.provider, s.score, k.confidence,
                    k.keyword_suffix
             FROM suggestions s
             JOIN prefix_keywords k ON k.suggestion_id = s.id
             WHERE k.keyword_prefix = :keyword_prefix AND
                   k.keyword_suffix LIKE :keyword_suffix || '%' AND
                   s.provider = :provider
             GROUP BY s.id
             ORDER BY rank DESC, s.score DESC
             LIMIT :suggestions_limit",
```

Reply:
Thanks so much, @0c0w3!

One more suggestion, @tiftran: SQLite's [query optimizer _can_ optimize `x%` `LIKE` patterns to be prefix matches](https://www.sqlite.org/optoverview.html#the_like_optimization), but unfortunately _won't_ by default—because `TEXT` columns are case-sensitive, but `LIKE` by default is case-insensitive. There's a pragma to change the runtime behavior of `LIKE`, but it's deprecated.

We could change the collation of the `keyword_prefix` and `keyword_suffix` columns in the schema, but there's an easier way. Remember that `BETWEEN` trick we used a while back?

I think we also want to flip the order of the terms in the `ORDER BY`, so that we sort by the suggestion's `score` first, and use the keyword's `rank` to break the tie.

Here's @0c0w3's suggestion, with those two updates:

```
SELECT s.id, MAX(k.rank) AS rank, s.title, s.url, s.provider, s.score, k.confidence,
       k.keyword_suffix
FROM suggestions s
JOIN prefix_keywords k ON k.suggestion_id = s.id
WHERE k.keyword_prefix = :keyword_prefix AND
      k.keyword_suffix BETWEEN :keyword_suffix AND :keyword_suffix || x'FFFF' AND
      s.provider = :provider
GROUP BY s.id
ORDER BY s.score DESC, rank DESC
LIMIT :suggestions_limit
```

And here's the query plan for it (with `LIMIT` omitted, to make sure we aren't fetching multiple results):

```
sqlite> .mode column
sqlite> .eqp on
sqlite> SELECT s.id, MAX(k.rank) AS rank, s.title, s.url, s.provider, s.score, k.confidence,
   ...>        k.keyword_suffix
   ...> FROM suggestions s
   ...> JOIN prefix_keywords k ON k.suggestion_id = s.id
   ...> WHERE k.keyword_prefix = 'soft' AND
   ...>       k.keyword_suffix BETWEEN 'li' AND 'li' || x'FFFF' AND
   ...>       s.provider = 4
   ...> GROUP BY s.id
   ...> ORDER BY s.score DESC, rank DESC;
QUERY PLAN
|--SEARCH k USING PRIMARY KEY (keyword_prefix=? AND keyword_suffix>? AND keyword_suffix<?)
|--SEARCH s USING INTEGER PRIMARY KEY (rowid=?)
|--USE TEMP B-TREE FOR GROUP BY
`--USE TEMP B-TREE FOR ORDER BY
id  rank  title                                                   url                                                           provider  score  confidence  keyword_suffix
--  ----  ------------------------------------------------------  ------------------------------------------------------------  --------  -----  ----------  --------------
2   1     ‘It’s Not Just Burnout:’ How Grind Culture Fails Women  https://getpocket.com/collections/its-not-just-burnout-how-g  4         0.25   0           living        
                                                                  rind-culture-failed-women 
```

Reply:
Thanks @linabutler, I believe the `ORDER BY` is performed after the grouping? Meaning `s.score` is actually a "bare column" and will be chosen in an undefined manner from one of the grouped suggestions. I glossed over that in my suggestion but it doesn't matter in practice since all suggestions of the same type have the same score, and although that might change someday, we don't have any plans to. (From the [Sqlite SELECT doc](https://sqlite.org/lang_select.html))

Reply:
Yep, you're totally right! I think that's OK, though, because we're grouping by `s.id`, and `s.id` is the primary key for the `suggestion` table? IIUC, that means we can have multiple rows for the same suggestion, but their `score`, `title`, `url`, and `provider` should all be identical—is that right?

Reply:
Sorry, I confused myself and my comment added nothing to this discussion! :-) You're right, we definitely want to sort first by score.

Reply:
> `k.rank` is unique for each `k.suggestion_id`

Actually, no, I'm wrong! `k.rank` is unique for `(k.confidence, k.suggestion_id)`, because Pocket has separate arrays of low-confidence and high-confidence keywords. So maybe we want to `GROUP BY s.id, k.confidence`?

Reply:
okay hold on, so we want s.id and k.confidence for pocket, and just s.id for amo?

Reply:
Not at all; thanks for thinking this through together!

How does the `k.confidence` addition look to you, @0c0w3 and @tiftran?

I'm having a hard time thinking of how that case could show up in practice—maybe in something like:

```
{
    "lowConfidenceKeywords": ["toxic work culture", "soft life", "work-life balance"],
    "highConfidenceKeywords": ["burnout women", "soft living"]
}
```

If your query is `soft li`, we have a low-confidence ("soft life", `confidence = 0`) and a high-confidence ("soft living", `confidence = 1`) with the same rank (1)—so we want to use the confidence and the rank to break the tie. And since we can `LIMIT` the number of suggestions we return, we want to sort low-confidence keywords first (`ASC`)—because if we sorted high-confidence first, and had `LIMIT 1`, we'd need an exact match for `soft living`, and so we wouldn't show the Pocket suggestion at all.

> okay hold on, so we want s.id and k.confidence for pocket, and just s.id for amo?

@tiftran Yep, I _think_ so! AMO (and MDN! I forgot about those 🙈 —they also use the `prefix_keywords` table) keywords don't have a confidence, so their `rank` should always be unique for `suggestion_id`.

Reply:
Looks good to me, I forgot about `confidence`.

Reply:
@tiftran and I talked through this a bit on Slack, and decided on:

```
...
GROUP BY s.id, k.confidence
ORDER BY s.score DESC, rank DESC
```

We group by `(id, confidence)` to catch the case where a suggestion has a high-confidence and a low-confidence keyword with the same first word and the same rank. But we order by just `(score, rank)`, and drop the `LIMIT`.

That way, the query will return a row for the high-confidence keyword, and another row for the low-confidence keyword. The `query_rows_and_then_cached` closure will be called for each row, which will determine which of the keywords matches based on the confidence—and drop the other.

Since `query_rows_and_then_cached` returns an iterator, we can use `.take()` to do the limit in Rust:

```rs
)?.into_iter().flatten().take(
    query.limit.and_then(|limit|
        usize::try_from(limit).ok()
    ).unwrap_or(usize::MAX)
).collect();
```

Reply:
feedback changes made. Lots of love and appreciation @bendk, @0c0w3 , @linabutler for helping me simplify and improve the sql 💜